alf/networks/networks.py

```diff
@@ -115,3 +118,122 @@ def __init__(self, block, input_tensor_spec=None, activation=torch.relu_):
     def forward(self, x, state=()):
         y, state = self._block(x, state)
         return self._activation(x + y), state
+
+
+class TemporalStack(Network):
+    """Stack features temporally.
+
+    Suppose input_size=(), stack_size=2, pooling_size=2, the following table
+    shows the output of different mode for an input sequence of 1,2,3,4,5 (ignoring
+    batch dimension)
+
+           1,        2,        3,        4,          5
+    skip:  [0, 1],   [0, 1],   [1, 3],   [1, 3],     [3, 5]
+    avg:   [0, 0],   [0, 1.5], [0, 1.5], [1.5, 3.5], [1.5, 3.5]
```

Comment:
should the third output be [1.5, 0] instead? (shifting happens after every pooling_size to be consistent with 'skip' mode). This updating rule is well defined or defined by ourselves?

Reply:
Yeah I know how different modes update, but then there is inconsistency in their outputs. As for 'avg' and 'max' can only be obtained every pooling_size steps, we can always pad the last value with the second last value when shifting. Suppose now the pooling size is large=5, then for the current 'max', it will be something like:

```
[0,0], [0,0], [0,0], [0,0], [0,5], |   [0,5], [0,5],[0,5],[0,5],[5,10], | ....
```
Another way of updating is 
```
[0,0], [0,0],[0,0],[0,0],[0,5] | [5,5], [5,5], [5,5], [5,5], [5,10], | ....
```
The second output will have less delay in the first entry of the output window, and is more consistent with 'skip'. I think it's depending on what kind of desired output we want.

Reply:
This will make the shift unnatural. It needs to have a special handling at step pooling_size. And in terms of information contained in the output, they contains same information.

Reply:
I just felt that the first entry in this case will have a delay of pooling_size - 1. But yeah, the two sequences have the same information at N*pooling_size.